hydra-node/src/Hydra/SqlLitePersistence.hs

```diff
@@ -0,0 +1,323 @@
+-- | Handles to save/load files across the hydra-node using sql-lite.
+-- We use a simple JSON encoding and two modes of operation to store things: Full and Incremental.
+module Hydra.SqlLitePersistence where
+
+import Hydra.Prelude
+
+import Conduit (ConduitT, ResourceT, yield)
+import Control.Concurrent.Class.MonadSTM (
+  newTMVarIO,
+  newTVarIO,
+  putTMVar,
+  readTVarIO,
+  takeTMVar,
+  writeTVar,
+ )
+import Data.Aeson qualified as Aeson
+import Data.Char (isDigit)
+import Database.SQLite.Simple (
+  Connection,
+  Only (..),
+  close,
+  execute,
+  execute_,
+  open,
+  query_,
+  withConnection,
+ )
+import System.Directory (createDirectoryIfMissing, getFileSize, removeFile)
+import System.FilePath (takeBaseName, takeDirectory, takeExtension)
+
+data PersistenceException
+  = PersistenceException String
+  | IncorrectAccessException String
+  deriving stock (Eq, Show)
+
+instance Exception PersistenceException
+
+-- | Handle to save and load files to/from db using JSON encoding.
+data Persistence a m = Persistence
+  { save :: ToJSON a => a -> m ()
+  , load :: FromJSON a => m (Maybe a)
+  , closeWriteDb :: m ()
+  , closeReadDb :: m ()
+  , closeDb :: m ()
+  , dropDb :: m ()
+  }
+
+createWriteConnectionV :: FilePath -> IO (TMVar IO Connection)
+createWriteConnectionV fp = do
+  createDirectoryIfMissing True $ takeDirectory fp
+  conn <- open fp
+  execute_ conn "PRAGMA journal_mode = WAL;"
```

Comment:
I remember when working on this we had problems with performance, hopefully these settings make the sqlitedb faster to read/write.

Reply:
the increase in performance comes from using a singleton connection rather than opening a new one for every db action